graphs/matching: implement the Micali–Vazirani maximum cardinality matching algorithm

janmenjayap opened the pull request:
<!-- ^ Please provide a concise and informative title. -->
The objective of this PR is to implement the `𝒪(|E|⋅√|V|)` Micali- Vazirani matching Algorithm.

<!-- ^ Don't put issue numbers in the title, do this in the PR description below. -->
<!-- ^ For example, instead of "Fixes #12345" use "Introduce new method to calculate 1 + 2". -->
<!-- v Describe your changes below in detail. -->
<!-- v Why is this change required? What problem does it solve? -->
<!-- v If this PR resolves an open issue, please link to it here. For example, "Fixes #12345". -->
Fixes #40091.


### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->

- [x] The title is concise and informative.
- [x] The description explains in detail what this PR is about.
- [x] I have linked a relevant issue or discussion.
- [x] I have created tests covering the changes.
- [x] I have updated the documentation and checked the documentation preview.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on. For example, -->
Nothing as of now (up to my knowledge).

<!-- - #12345: short description why this is a dependency -->
<!-- - #34567: ... -->

cc: @dcoudert, @dimpase.

janmenjayap committed:
introduced micali-vazirani matching algorithm

janmenjayap committed:
Merge branch 'develop' into micali_vazirani

janmenjayap committed:
added micali-vazirani matching algorithm

janmenjayap committed:
updated the code style and fixed a bug

janmenjayap committed:
updated the code style

janmenjayap committed:
removed micali_vazirani_matching.py

janmenjayap committed:
implemented micali-vazirani matching

janmenjayap committed:
updated the tenacity_bridges_map

dcoudert commented:
You have several failing doctests and the code structure and documentation must be seriously improved.

janmenjayap committed:
Merge branch 'develop' into micali_vazirani

dcoudert commented:
We still have several failing doctests and the code structure and documentation must be seriously improved. You should either make a dedicated function for this algorithm or a class if it eases structuring the code.

janmenjayap committed:
Merge branch 'develop' into micali_vazirani

janmenjayap committed:
updated the initialization; added some comments

janmenjayap committed:
updated doctests

janmenjayap committed:
updated doctests

janmenjayap committed:
updated doctests

janmenjayap committed:
updated doctests

janmenjayap committed:
refactored the dicts to lists

janmenjayap committed:
updated compute_initial_maximal_matching()

janmenjayap committed:
updated the code

janmenjayap committed:
removed unnecessary blank lines

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1195,16 +1213,44 @@ def matching(G, value_only=False, algorithm='Edmonds',
+        2
+        sage: g = graphs.SylvesterGraph()
+        sage: m = g.matching(algorithm='Micali-Vazirani')
+        sage: n = g.matching(algorithm='Edmonds')                                   # needs sage.networkx
```
the tag should be `# needs networkx`.
With the wrong tag, the command is not run, and so `n` is not defined.

janmenjayap replied:
That's interesting.
I was not aware of that. Thanks!

Originally, I found some other lines (wrt `Edmonds` algorithm, aka the default one for the method `matching()`) with the tag `# needs sage.networkx` in the same list of doctests: https://github.com/sagemath/sage/blob/develop/src/sage/graphs/matching.py#L1192.

dcoudert reviewed (changes requested):
Please make a specific method for the new algorithm. It is way too long for being included inside the `matching` method.

janmenjayap committed:
Updated the tag: # needs networkx

dcoudert reviewed (changes requested):
a first round of comments. This is a giant PR, not easy to review at once.

janmenjayap commented on src/sage/graphs/matching.py:
```diff
@@ -1273,7 +1319,752 @@ def weight(x):
+                        buckets[degree].add(vertex)
+
+                # Find the initial minimum positive degree present in the buckets
+                minimum_degree = 1
```
Yes, the graph `J` does not have any isolated vertex, because before calling `compute_initial_maximal_matching()`, we get rid of all isolated vertices while constructing `H` from `G`.

```
            # relabel vertices of G to 0..n‑1
            H = G.copy()
            N = H.order()

            for vertex in H:
                if not H.degree(vertex):
                    H.delete_vertex(vertex)
                    continue

            vertex_to_index_map = H.relabel(inplace=True, return_map=True)
            index_to_vertex_map = [None] * N
            for vertex, index in vertex_to_index_map.items():
                index_to_vertex_map[index] = vertex
```

janmenjayap replied:
If `J` has isolated vertex, that does not compute any edge to the initial matching.

janmenjayap committed:
updated the code and the comments

janmenjayap commented:
I will refactor the code and push it by converting it into a class structure outside the `matching()` method, shortly.

janmenjayap committed:
Merge branch 'develop' into micali_vazirani

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1366,91 +1366,79 @@ def compute_initial_maximal_matching() -> None:
-                                if new_degree > 0:
-                                    buckets[new_degree].add(vertex)
+                    # Remove these vertices from the bucket lists and update the degree in degree list
+                    for vertex in (u, v):
```
vertex `u` has been obtained with a pop operation. So it is no longer in the bucket. However, its degree must be updated.

janmenjayap committed:
added :class:

janmenjayap committed:
merged branch 'micali_vazirani' of https://github.com/janmenjayap/sage into micali_vazirani

janmenjayap committed:
merged branch 'micali_vazirani' of https://github.com/janmenjayap/sage into micali_vazirani

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1366,91 +1366,79 @@ def compute_initial_maximal_matching() -> None:
-                degree_map: dict[Hashable, int] = {vertex: J.degree(vertex) for vertex in J}
-                maximum_degree = max(degree_map.values()) if degree_map else 0
+                # Get the current maximum degree in J
+                initial_degree_sequence = J.degree_sequence()
```
Observe that `degree_sequence` returns the list of degrees sorted by decreasing values. Consequently, `initial_degree_sequence[0]` is the maximum degree and `initial_degree_sequence[-1]` is the minimum degree.
Reading the documentation of a method is useful.

The method takes time $O(n\log{n})$, while a `min` or `max` operation on the list of degrees takes time $O(n)$.

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1366,91 +1366,79 @@ def compute_initial_maximal_matching() -> None:
-                    minimum_degree += 1
+                # buckets[d] contains the set of vertices currently having degree d in J
+                buckets: List[set] = [set() for _ in range(maximum_degree + 1)]
+                for v in J:
```
why not 
```py
for v, d in enumerate(degree):
    buckets[d].add(v)
```

janmenjayap commented:
I have added the required class (`MicaliVaziraniMatching`).

Please note that the examples added, it computes the maximum matching in the initial stage of maximal matching computation itself. I will be adding more examples (on possibly larger graphs) for this class.

Also, I will be adding another method under the class to output the statistics of the algorithm flow at the end (for instance, the total number of phases, what are the bridges encountered during each phase, what are the final min/ max level of each vertex, etc.)

janmenjayap committed:
merged branch 'micali_vazirani' of https://github.com/janmenjayap/sage into micali_vazirani

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1627,3 +1684,704 @@ def M_alternating_even_mark(G, vertex, matching):
+
+class MicaliVaziraniMatching:
+    r"""
+    Class for computing maximum cardinality matching in a simple undirected
```
improve the 1 line description, add description what this class do, possibly a reference to a paper and some examples.

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1627,3 +1684,704 @@ def M_alternating_even_mark(G, vertex, matching):
+        base: int
+        peaks: Tuple[int, int]
+
+    def __init__(self, G: Graph) -> None:
```
all methods, including `__init__` must have a 1 line description, a description of the input and some doctests.

janmenjayap committed:
updated compuute_initial_maximal_matching()

janmenjayap committed:
updated compute_initial_maximal_matching()

janmenjayap committed:
PEP 8 fix

janmenjayap committed:
PEP 8 fix

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1627,3 +1684,704 @@ def M_alternating_even_mark(G, vertex, matching):
+                self.G.delete_vertex(vertex)
+
+        # indexing the vertices
+        self.vertex_to_index = self.G.relabel(inplace=True, return_map=True)
```
a more classical form
```py
self.index_to_vertex = list(G)
self.vertex_to_index = {u: I for I, u in enumerate(self.index_to_vertex)}
self.G.relabel(perm=self.vertex_to_index, inplace=True)
```

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1627,3 +1684,704 @@ def M_alternating_even_mark(G, vertex, matching):
+        self.successor: List[List[int]] = [[] for _ in range(self.N)]
+        self.color: List[int] = [None] * self.N
+        self.search_level_vertices: List[int] = list(range(self.N))
+        self.edge_scanned: Dict[int, int] = {self.edge_to_index(u, v): -1 for (u, v, _) in self.G.edge_iterator()}
```
```py
        self.edge_scanned: Dict[int, int] = {self.edge_to_index(u, v): -1 for u, v in self.G.edge_iterator(labels=False)}
```

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1627,3 +1684,704 @@ def M_alternating_even_mark(G, vertex, matching):
+        self.INFINITY = float('inf')
+
+    # indexing the edges
+    def edge_to_index(self, i: int, j: int) -> int:
```
why not using mappings ?
```py
self.index_to_edge = list(self.G.edges(labels=False, sort_vertices=True))
self._edge_to_index = {e: I for I, e in enumerate(index_to_edge)}

def edge_to_index(self, i: int, j: int) -> int:
    if I > j:
        return self. _edge_to_index[j, I]
    return self. _edge_to_index[I, j]
```

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1627,3 +1684,704 @@ def M_alternating_even_mark(G, vertex, matching):
+        J: Graph = self.G.copy(immutable=False)
+
+        # Get the degree sequence of J
+        degree_sequence = J.degree_sequence()
```
as previously said, `degree_sequence` is the sorted list of degrees by decreasing values. No need to use `min` and `max` on a sorted list.

You should first build the list of degrees, and then take the min and max values.

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1627,3 +1684,704 @@ def M_alternating_even_mark(G, vertex, matching):
+                edge_index = self.edge_to_index(u, v)
+                l = self.G.edge_label(u, v)
+
+                if self.edge_scanned[edge_index] != self.phase_index and self.M.has_edge(u, v, l) == parity and self.deletion_phase[v] != self.phase_index:
```
put on multiple lines to avoid long lines.

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1627,3 +1684,704 @@ def M_alternating_even_mark(G, vertex, matching):
+                        if self.M.size() == self.G.order() // 2:
+                            return is_augmented
+
+            else:
```
use `elif`

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1627,3 +1684,704 @@ def M_alternating_even_mark(G, vertex, matching):
+
+        # Set the starting point for each of red and green DFS's
+        red_stack, green_stack = [], []  # Stack saves previously traversed vertices
+        red_vertex, green_vertex = self.get_bud(source_red_vertex), self.get_bud(source_green_vertex)  # Set the initial point for both DFS's
```
please don't put comments at the end of the lines. The lines are very long. The comments can come before the instruction.

dcoudert commented on src/sage/graphs/matching.py:
```diff
@@ -1627,3 +1684,704 @@ def M_alternating_even_mark(G, vertex, matching):
+
+        # Boolean variables are initiated
+        no_augmentation_found = False if not self.min_level[red_vertex] and not self.min_level[green_vertex] else True
+        collision = True if red_vertex == green_vertex else False
```
simply `collision = red_vertex == green_vertex`

dcoudert reviewed (commented):
another set of comments.
Not easy to review...

janmenjayap committed:
fixed circular import of sage.graphs.graph.Graph

janmenjayap committed:
fixed circular import of sage.graphs.graph.Graph

janmenjayap committed:
refactor: code improvements

janmenjayap committed:
refactor: code improvements

dcoudert commented:
any update on fixing the bugs ?

janmenjayap committed:
refactor: resolve linting errors and code style issues in matching.py

janmenjayap committed:
Merge branch 'develop' into micali_vazirani

janmenjayap committed:
fix: resolve NameError in petal helpers (missing self.)

janmenjayap committed:
docs: document MicaliVaziraniMatching class and add references

janmenjayap committed:
fix: correct construction and bridge bucketing in MicaliVaziraniMatching

- relabel vertices in a single pass and drop isolated vertices before setting N, so construction works for arbitrary vertex labels
- build the edge index before edge_scanned, which depends on it
- file a bridge in label_max only when its tenacity is finite
- remove a stray quote from two error messages

janmenjayap committed:
docs: document MicaliVaziraniMatching.__init__

janmenjayap committed:
docs: document MicaliVaziraniMatching.edge_to_index

janmenjayap committed:
docs: document compute_initial_maximal_matching and fix its return annotation

janmenjayap committed:
docs: document MicaliVaziraniMatching.start_new_phase

janmenjayap committed:
docs: document MicaliVaziraniMatching.get_bud

janmenjayap committed:
docs: document MicaliVaziraniMatching.MIN

janmenjayap committed:
docs: document MicaliVaziraniMatching.MAX and label_max

janmenjayap committed:
docs: document MicaliVaziraniMatching DDFS, advance_DFS and reverse_DFS

janmenjayap committed:
docs: document MicaliVaziraniMatching form_blossom and form_petal

janmenjayap committed:
docs: document MicaliVaziraniMatching unfold_petal and unfold_path_in_petal

janmenjayap committed:
docs: document MicaliVaziraniMatching augment and get_path

janmenjayap committed:
docs: document MicaliVaziraniMatching search and get_matching

janmenjayap committed:
style: move end-of-line comments to their own line

dcoudert reviewed (commented):
a few comments. I have not tested the method yet.

janmenjayap committed:
style: wrap long signatures/lines and address review nits in matching.py

Wrap all over-length method signatures (DDFS, advance_DFS, reverse_DFS, form_blossom, form_petal, unfold_path_in_petal, augment) and split genuinely long code and comment lines for readability, with no variable renaming and no new named locals.

Also address review comments:
- use edge_to_index[i, j] indexing form
- guard with `if self.M.degree(u):`
- reorder the MIN-phase early return
- spell "neighbor" (prose and identifier) consistently

janmenjayap committed:
graphs/matching: simplify MAX return and DDFS boolean init

Both rewrites are behavior-preserving:

MAX: the tail always returns False — the only truthy exit is the early
return taken when the matching becomes perfect. Drop the dead
is_augmented = False and return False directly, keeping the
num_augmentations increment.

DDFS: False if not A and not B else True is De Morgan's law — it
equals bool(A or B). Use bool(...) so the flag stays strictly boolean
(it is later reassigned = False).

janmenjayap committed:
graphs/matching: flatten branch structure in unfold_path_in_petal

janmenjayap committed:
graphs/matching: simplify greedy seed loop in compute_initial_maximal_matching

All three changes preserve the matching size — only the greedy seed's
iteration order changes, never the final matching:

- Fold the J.order() <= 1 break into the while condition.
- Test isolation via the maintained degree[u] (invariant-equal to
  J.degree(u)) instead of building a neighbor list just to check
  emptiness; drop the now-redundant list(...) wrapper.
- After each match, purge vertices that just became isolated and resume
  the next scan from degree 1 — they can never be matched.

Add TESTS doctests cross-checking the seed's maximality and the final
matching size against Edmonds' algorithm, including a long-time random
sweep that exercises the new purge.

janmenjayap committed:
graphs/matching: accept non-simple input in MicaliVaziraniMatching

Previously the constructor rejected graphs with loops or multiple edges
with a ValueError. Instead, accept them and simplify internally: a
maximum matching of to_simple(G) is a maximum matching of G, since a
loop can never be matched and parallel edges are redundant for the
unweighted matching computed here.

- __init__: route loop/multiedge-allowing graphs through to_simple()
  (which also clears the loops/multiedge flags so downstream mutation of
  self.G is safe); copy otherwise. Drop isolated vertices via
  degree_iterator(labels=True).
- matching(): drop the now-redundant to_simple() at the call site.
- Update the class and constructor docstrings, flip the TESTS doctest
  that expected a ValueError on a loop graph, and add doctests for
  Micali-Vazirani on a loop/multigraph.

dcoudert commented:
I did some tests and the new algorithm is working well and is much faster than other methods.
However, I raised an issue with `Edmonds`. You should check
```py
sage: def BA(n):
....:     B = graphs.RandomBarabasiAlbert(n, 2)
....:     G = Graph(B, multiedges=True, loops=True)
....:     G.add_edges(B.edges())
....:     G.add_edges((u, u) for u in G)
....:     return G
....: 
sage: G = BA(1000)
sage: G.matching(value_only=True, algorithm="Micali-Vazirani")
418
sage: G.matching(value_only=True, algorithm="Edmonds")
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[60], line 1
----> 1 get_ipython().run_line_magic('time', 'G.matching(value_only=True, algorithm="Edmonds")')

File <timed eval>:1
----> 1 'Could not get source, probably due dynamically evaluated source code.'

File ~/sage/src/sage/graphs/matching.py:1295, in matching(G, value_only, algorithm, use_edge_labels, solver, verbose, integrality_tolerance)
   1293         g.add_edge(u, v, weight=w)
   1294 else:
-> 1295     for u, v in L:
   1296         g.add_edge(u, v)
   1297 d = networkx.max_weight_matching(g)

ValueError: not enough values to unpack (expected 2, got 1)
```

janmenjayap commented:
Confirmed: this is a pre-existing bug in `Graph.matching()`, not in the Micali–Vazirani code. The self-loop filter uses an identity test (if u is v), which only matches loops whose endpoints are the same Python object. CPython caches integers −5..256, so it works for small labels but lets a loop on a higher-labelled vertex through; its `frozenset((u, u))` then collapses to a single element and fails to unpack at `for u, v in L`. It affects the Edmonds and LP branches; Micali–Vazirani is unaffected because it filters loops differently.

The `BA(1000)` graph triggers it because the `(u, u)` loops land on vertices labelled above 256. Minimal repro:

```python
  sage: G = Graph([(300, 300), (300, 301)], loops=True)
  sage: G.matching(value_only=True, algorithm="Edmonds")
  ...
  ValueError: not enough values to unpack (expected 2, got 1)
```
  
The fix is to compare endpoints by value `(u == v)`. Since it's a separate, pre-existing issue, I have opened it as its own PR rather than folding it into this one: #42348 (as this is unrelated to Micali-Vazirani matching). It includes a regression doctest on a self-loop with a label above 256.

Could you please take a look at it when you get a chance?
Thank you

dcoudert commented:
I think we can now make this algorithm the default one

janmenjayap commented:
#### Audit of the wider `is`-on-vertices pattern in `src/sage/graphs/`
I swept the whole graphs module for identity comparisons of vertex/edge endpoints. There are five such sites; only two are real bugs.

| Site | Code | Verdict |
|------|------|---------|
| `graph.py` `chromatic_symmetric_function` | `if u is not v` | **bug — fixed in #42369** |
| `matching.py` `matching` | `if u is v` | bug — already fixed in #42348 |
| `graph.py` `cores` | `v = y if x is u else x` | safe |
| `generic_graph.py` `has_multiple_edges` | `if a is u / elif b is u` | safe |
| `bipartite_graph.py` (bipartition helper) | `vertex_to_component[u] is vertex_to_component[v]` | safe — not a vertex comparison |


Could you please take a look the new PR when you get a chance?
Thank you

janmenjayap committed:
Merge branch 'develop' into micali_vazirani

janmenjayap committed:
refactor(matching): represent the MV matching as a mate[] array

Replace the auxiliary Graph (self.M) that stored the partial matching
with self.mate, the integer mate-array standard in the matching
literature: mate[v] is the vertex matched to v, or EXPOSED (-1). A
matching_size counter replaces self.M.size(), and is_exposed /
is_saturated predicates name the exposure test.

The per-edge "is (u, v) matched?" and per-vertex "is v matched?" queries
become O(1) array reads, label bookkeeping is removed from the search
hot paths, and get_matching reconstructs the edge set -- with labels --
once at the boundary from mate and index_to_vertex. The returned
matching and its cardinality are unchanged.

dcoudert commented:
I did some test and it's very good
```py
sage: G = graphs.RandomBarabasiAlbert(10000, 2)
sage: %time G.has_perfect_matching(algorithm='Edmonds')
CPU times: user 11.3 s, sys: 74.1 ms, total: 11.3 s
Wall time: 11.4 s
False
sage: %time G.has_perfect_matching(algorithm='Micali-Vazirani')
CPU times: user 182 ms, sys: 3.55 ms, total: 185 ms
Wall time: 184 ms
False
sage: %time G.is_bicritical(algorithm='Edmonds')
CPU times: user 11.3 s, sys: 58.1 ms, total: 11.4 s
Wall time: 11.4 s
False
sage: %time G.is_bicritical(algorithm='Micali-Vazirani')
CPU times: user 185 ms, sys: 2.34 ms, total: 187 ms
Wall time: 186 ms
False
sage: %time M = G.matching(algorithm='Edmonds')
CPU times: user 10.9 s, sys: 28.7 ms, total: 10.9 s
Wall time: 10.9 s
sage: len(M)
4214
sage: %time M = G.matching(algorithm='LP')
CPU times: user 1.62 s, sys: 28.6 ms, total: 1.65 s
Wall time: 1.58 s
sage: len(M)
4214
sage: %time M = G.matching(algorithm='Micali-Vazirani')
CPU times: user 174 ms, sys: 1.69 ms, total: 176 ms
Wall time: 175 ms
sage: len(M)
4214
```
We should make it the default algorithm for undirected unweighted graphs. This could be done by using `algorithm=None` and in this case select the most suitable algorithm.

janmenjayap committed:
graphs/matching: make Micali-Vazirani the default for unweighted matching

Micali-Vazirani computes an unweighted maximum matching in O(sqrt(|V|).|E|)
and, unlike Edmonds' algorithm, needs no optional NetworkX dependency, so it
is a better default for the common unweighted case.

- matching(): resolve algorithm=None to Micali-Vazirani for an unweighted
  matching and to Edmonds for a weighted one (use_edge_labels=True)
- matching(): return an Integer when value_only=True under Micali-Vazirani
- has_perfect_matching, is_bicritical, is_factor_critical, is_matching_covered:
  default algorithm to None and delegate the choice to the graph's own
  matching(), so BipartiteGraph keeps using Hopcroft-Karp instead of rejecting
  the request
- MicaliVaziraniMatching: build the working graph via the base Graph
  constructor so it can mutate freely, instead of dispatching into a subclass's
  guarded mutators (e.g. MatchingCoveredGraph.delete_vertices)
- matching_covered_graph: assert matching invariants in the affected doctests
  rather than one specific matching; fix the malformed subdivide_edges assert
- docs: update the matching() ALGORITHM section and drop stale "# needs
  networkx" tags from default-matching doctests

dcoudert commented:
Do you have further improvements in minds or can I finalize the review ?

janmenjayap commented:
#### Micali–Vazirani Algorithm: to-dos

**Status:** passes the full global doctest suite (2026-06-16). Currently runs at 61× faster than Edmonds' and at 9.03× faster than LP on random large graphs.

<hr/>
Here's a to-do list from my end, that are remaining to be done. I would appreciate if you could please go through this and possibly suggest any changes, when you get a chance

Thank you

<hr/>

#### A. More robust test cases
- [x] Add a helper asserting each result is a *valid* matching (edges exist, no vertex covered twice) **and** *maximum* (size == Edmonds), incl. a randomized sweep (today's random sweep checks size only).
- [x] Add complex/ adversarially constructed graphs that actually force multiple augmentation phases / blossom formation / nested blossoms. Don't guess from shape: verify by re-running coverage and keep the graphs that move the uncovered set below.

#### B. Docs
- [x] Rewrite class docstring for the extended-phase (HS2017) variant; complexity O(√V·E·log V) (path-compression-only union-find), not the α(E,V) bound (needs union-by-rank) nor the optimal O(√V·E) (needs Gabow's set-merging): both unimplemented.
- [x] B2. *(only if B1 cites it)* add `[Vaz2020]` to the reference index. **Do not add `[Gabow2014]`.** (`MV1980`/`Vaz1994`/`HS2017` already present.)
- [x] B3. "an 1-regular" -> "a 1-regular".

#### C. Cleanup and modernization
- [x] C1. Builtin generic annotations (`List` -> `list`, `Optional[X]` ->`X | None`).
- [x] C2. Remove the two dead `tenacity_bridges_map.extend(...)` branches (lines 2347 `MIN`, 2479 `label_max`).

#### D. Correctness (re-derive against the `mate[]` representation)
- [x] D1. Make `get_bud` iterative (path-compressed find).
- [x] D2. Restore augment's path-validation guard (length + strict alternation).
- [x] D3. Recursion/stack guard for nested-blossom unfolding.

#### E. Refactor
- [x] E1. Collapse `level`/`min_level`/`max_level` into `evenlevel`/`oddlevel`.
- [x] Replace the `float('inf')` sentinel (`self.INFINITY = float('inf')`) with the **integer** sentinel `INFINITY = 2*N + 2`.

#### F. Coverage gaps (committed HEAD)
Uncovered lines : `60, 181-184, 188-190, 443, 450-470, 481, 502-503, 670, 1023, 1028, 1050, 1423, 1543, 2171-2173, 2319, 2347, 2425, 2479, 2572-2573, 2577-2579, 2623, 2629, 2873-2901, 2934-3015, 3062, 3090-3095, 3139, 3147-3153, 3199`

- [ ] F1. Add tests for the reachable MV lines (covered by Phase A once the right graphs are found): `2171-2173` (greedy isolated-vertex drop), `2319` MIN, `2425` MAX, `2572-2579`/`2623`/`2629` DDFS, `2873-2901` `unfold_petal`, `2934-3015` `unfold_path_in_petal`, `3062`/`3090-3095` `augment`, `3139`/`3147-3153` `get_path`, `3199` `search`. (`1423` is the LP branch --- optional `# needs sage.numerical.mip` test.)
- [x] F2. `2347`, `2479` --- disappear with C2.
- [x] F3. `60` --- `TYPE_CHECKING`-only import; annotate as unreachable.
- [ ] F4. `181-190`, `443-503`, `670`, `1023-1050`, `1543` --- pre-existing gaps in
        non-MV sibling functions; out of scope.

#### G. Optional
- [x] G.1 Fully de-recurse `unfold_petal`/`unfold_path_in_petal` (explicit stack)

dcoudert commented:
I'm not sure what you are asking for. I think you have a much better understanding of this algorithm and code than me, so if you think some improvements are needed / useful / necessary, feel free to do them. 
So improvements can also be postponed to future PR if there is no urgent need for them (e.g., typing).

janmenjayap commented:
Thank you, that makes sense; and I appreciate the flexibility on what can move to a follow-up PR.

My plan is to do most of these now. I have started with the easy/ minor changes (the cleanup, docs, and the safe correctness fixes), and I am working through those while thinking about the more complex test cases in parallel: the adversarial graphs that actually force multiple augmentation phases and (nested) blossom formation take some care to construct, so I would like to get those right rather than rush them.

The typing modernization is quick and low-risk, so I will likely just fold it into this PR too rather than defer it; but good to know I have the option to push it to a follow-up if it starts to grow.

I will push the changes incrementally so they are easy to review.
Thanks again!

janmenjayap committed:
docs: clarify Micali-Vazirani variant and complexity, fix grammar

- Describe the implemented algorithm as the extended search-phase
  variant of Huang and Stein [HS2017].
- Replace the bald O(sqrt(V) E) claim with the recognized bound
  O(sqrt(V) E alpha(E, V)), where alpha is the inverse Ackermann
  function, arising from the path-compressed union-find that maintains
  blossom bases; note that the optimal O(sqrt(V) E) bound, requiring
  Gabow's incremental-tree set-merging, is not implemented.
- Link the inverse Ackermann function to its Wikipedia entry.
- Cite Vazirani's complete correctness proof [Vaz2020] and add it to the
  reference index (arXiv:2012.03582).
- Fix module-header grammar: "an 1-regular" -> "a 1-regular" (twice) and
  "if it the" -> "if the".

janmenjayap committed:
refactor(matching): tidy tenacity-bucket handling with an integer sentinel

- Use an integer "infinity" sentinel `INFINITY = 2*N + 2` (the length of
  `tenacity_bridges_map`) instead of `float('inf')`, keeping the level
  arrays homogeneously `int` and tightening their annotations.
- The guard `tenacity < self.INFINITY` now reads exactly as "this tenacity
  indexes a real bucket", so a bridge is filed only when it fits.
- Drop the two now-redundant `if tenacity >= len(self.tenacity_bridges_map)`
  auto-extend branches in `MIN` and `label_max`: they were the complement of
  that guard and never executed, and they were a latent cross-phase leak
  (the per-phase reset clears only a fixed index range).

janmenjayap committed:
style(matching): use builtin generic type annotations

- Replace `List`/`Dict`/`Set`/`Tuple` with builtin `list`/`dict`/`set`/`tuple` and `Optional[int]` with `int | None`; drop `Optional` from the typing import.
- These names were never imported and survived only via `from __future__ import annotations`; builtin generics need no import, so the latent `NameError` risk is removed.
- Keep the `Any` import (used at runtime in `Edge = tuple[int, int, Any]`).
- Correct the `color` annotation to `list[int | None]` (it holds both vertex colours and `None`).
- Annotate the empty-collection locals `next_search_level_vertices`, `erase_vertex_list`, and `path` as `list[int]`, where the element type is otherwise invisible.

janmenjayap committed:
graphs/matching: add validity + maximum-cardinality matching tests

Strengthen the matching() and MicaliVaziraniMatching doctests to assert
that results are valid (every matched edge exists, no vertex covered
twice) and of maximum cardinality (size == algorithm='Edmonds'), rather
than checking size alone.

- matching(): add a vertex-label-kind round-trip sweep (strings, int
  tuples, Sage integers, rationals) proving the relabel round-trip and
  arbitrary edge-label survival, an isolated-vertex-drop example, and a
  labelled blossom-on-a-stem example.
- compute_initial_maximal_matching(): add a valid_max_matching helper
  asserting validity and maximum cardinality, and route both the
  fixed-graph check and the randomized sweep through it (upgrading the
  previous size-only random sweep).
- All assertions are property-based (size, has_edge, label equality,
  no double-cover), so they are independent of which maximum matching MV
  returns.

janmenjayap committed:
graphs/matching: cover blossom/petal paths with adversarial test graphs

Drive the Micali-Vazirani augmentation machinery's doctest coverage from
82% to 91% with coverage-guided regression graphs.

- get_matching(): add a property-based regression block over three
  edge-induced graphs (orders 12/16/18), found by a coverage-guided
  search and each verified valid + maximum-cardinality (size ==
  algorithm='Edmonds'). They exercise the reachable cores of DDFS,
  blossom formation, unfold_petal, unfold_path_in_petal, get_path's
  petal branch, augment's erase BFS, and search; one of them also
  unfolds a nested petal.
- matching(): add an LP weighted value-only doctest
  (# needs sage.numerical.mip).
- Annotate the TYPE_CHECKING-only import as runtime-unreachable by
  design.
- compute_initial_maximal_matching(): per review, simplify the seed
  validity/maximum random check to a single RandomGNP instance and drop
  set_random_seed (the doctest framework varies the seed itself).
- The remaining uncovered MV lines are defensive reconstruction-failure
  guards, nested-blossom recursion not reached by inputs, and a dead
  greedy isolated-vertex branch.

janmenjayap committed:
refactor(matching): make get_bud iterative; document union-find complexity

Rewrite get_bud as an iterative two-pass union-find find (locate the bud,
then path-compress) instead of recursive, removing a stack-overflow source
on deeply nested blossoms. The rewrite is equivalent to the recursive
version -- same return value and same full path compression -- verified
over tens of thousands of random union-find forests; a new regression
doctest resolves a 10000-long bud chain that the recursive version could
not.

- Union-by-rank is deliberately not added: the blossom base must stay the
  set representative, so the union links by base, not rank.
- Correct the class-docstring time complexity from
  O(sqrt(V).E.alpha(E,V)) to the honest O(sqrt(V).E.log V)
  (path compression alone).
- Note that the alpha bound would require decoupling a rank-balanced
  forest from the bud labels, which is not implemented.
- Add the O(V+E) space-complexity statement.

janmenjayap committed:
fix(matching): validate reconstructed augmenting paths in augment

Before toggling, augment now checks that the reconstructed path is a
well-formed augmenting path of the current matching via a new
_is_valid_augmenting_path helper: even length (odd number of edges),
distinct vertices, both endpoints exposed, every consecutive pair a real
edge, and strict unmatched/matched alternation. A malformed reconstruction
returns False (leaving the matching untouched) instead of corrupting it.

- The check encodes augment's existing preconditions, so it is transparent
  on correct runs (matchings unchanged).
- The helper is covered directly by doctests for the accepting case and
  each rejecting case.

janmenjayap committed:
refactor(matching): de-recurse blossom unfolding via a generator trampoline

- Replace the mutually recursive unfold_petal and unfold_path_in_petal
  implementation with a generator trampoline so deeply nested blossoms no
  longer risk overflowing Python's recursion stack.
- Drive each public method through _run_trampoline, where every former
  recursive call becomes a yield of a sub-generator and the call stack lives
  on the heap.
- Keep the generator bodies otherwise line-for-line equivalent to the
  recursive versions.
- With get_bud already iterative, make the matching algorithm entirely
  recursion-free, so no recursion-limit guard is needed.
- Verify the refactor as matching-identical to the recursive implementation
  across ~8000 random and structured graphs, with the full doctest suite
  passing.
- Rename get_path to find_path: the method searches for and reconstructs an
  augmenting path, may return an empty list when reconstruction is not
  possible, and matches MV1980's FINDPATH naming alongside MIN/MAX/DDFS.

janmenjayap committed:
refactor(matching): store levels as even_level/odd_level only

- Replace the per-vertex `level` pair-lists and cached `min_level`/`max_level` arrays with the papers' two primitive arrays, `even_level` and `odd_level`.
- Derive `min_level` and `max_level` on demand via new `min_level(v)` and `max_level(v)` methods.

janmenjayap committed:
graphs/matching: doctest petal unfolding via an unseeded triangle chain

- Strengthen get_matching's TESTS block to exercise the petal-unfolding trampoline from two angles, sharing one validator.
- Refactor the is_valid_maximum_matching doctest helper to take the matching as an argument, reused by the blossom-graph check and the new triangle check.
- Document that the blossom graphs' greedy seed is non-empty but one short of maximum, so unfolding happens with the seed in place.
- Add a triangle_chain generator and get_matching_from_empty helper, with the greedy seed neutralised on the instance.
- Demonstrate that the seed is genuinely empty, then check valid + maximum matchings against Edmonds for chain lengths 1..20 plus a long-time deep case at length 60.
- Build the triangle-chain matching from empty because the greedy seed already solves those chains, ensuring augmentation routes through the blossoms.

janmenjayap committed:
docs(matching): use ``min_level`` etc. so underscores aren't parsed as LaTeX

janmenjayap committed:
graphs/matching: doctest label-shuffle invariance of the MV matching

- Add a get_matching TESTS doctest that applies several random vertex-label permutations to the regression blossom graphs and triangle chains.
- Run the triangle chains without the greedy seed, forcing augmentation through the blossoms.
- Assert every shuffled instance still yields a valid maximum matching.
- Guard against order-dependence bugs in bridge bucketing or double DFS, since internal indices are assigned in vertex-iteration order and a single fixed labelling could hide such issues.

janmenjayap committed:
graphs/matching: doctest the MV matching on sparse near-threshold random graphs

janmenjayap committed:
graphs/matching: doctest nested blossoms, structured/random sweeps and worst-case seeds

janmenjayap committed:
graphs/matching: drop set_random_seed(0) from MV matching doctests

janmenjayap committed:
graphs/matching: doctest the MV matching on symmetric and regular graphs

janmenjayap committed:
graphs/matching: doctest toroidal grids and factor-critical/deficiency graphs

janmenjayap committed:
graphs/matching: doctest determinism and the EdgesView return contract

janmenjayap committed:
Merge branch 'develop' into micali_vazirani

janmenjayap committed:
graphs/matching: reconstruct blossom arcs with a joint backtracking search

- The blossom path-unfolding rebuilt a petal's two arcs greedily and
  independently, which could produce a non-simple path rejected by
  `_is_valid_augmenting_path` or strand the second arc, dropping a shortest
  augmenting path through the blossom; `get_matching` could then return a
  matching below maximum cardinality, and on some inputs the unfold looped
  indefinitely.
- Rewrite the unfold as a single joint search: a `visited` set threaded
  through the whole expansion keeps every reconstructed segment simple and lets
  the search backtrack over branched support.
- For an even-level vertex the two arcs are searched together: the vertex-arc
  enumerates candidates while a `continuation` reconstructs the bud-arc
  avoiding it, returned via `result_holder`, backtracking across the bridge
  whenever the arcs cannot be made vertex-disjoint, so a simple alternating path
  is found whenever one exists.
- Nested petals stay single-solution since each vertex belongs to at most one
  petal.
- Remove the now-unused greedy helpers `_can_reach_in_petal` and
  `_nested_petal_reaches` and the `prefer_new_petal` branching, and drop the
  write-only `visit_mark` state left in `DDFS`.
- Add two reproducer graph6 strings as regression doctests; verified against
  Edmonds' algorithm on 115k+ random and blossom-structured graphs, seeded and
  seed-suppressed, and via the file's doctests.

janmenjayap committed:
graphs/matching: use petal/bud terminology consistently

- Document the [Vaz2020] terminology and stop using petal/bud and
  blossom/base interchangeably: petal/bud are the algorithmic objects the
  code builds and manipulates, while blossom/base are the corresponding
  graph-theoretic notions.
- Add a terminology glossary to the class docstring and an arc note to
  `unfold_petal` making this distinction explicit.
- Rename `Petal.base` to `Petal.bud`, `form_blossom` to `form_petal`, and
  the old per-side `form_petal` helper to `_attach_petal_side`. Existing
  names that already follow the terminology, including `get_bud`,
  `vertex_bud_map`, `vertex_petal_map`, and `Petal`, are unchanged.
- Update prose throughout the file to use petal/bud for the contracted
  structures handled by the algorithm and reserve blossom/base for the
  graph-theoretic notions.
- No behaviour change; the file's doctests and differential fuzz tests
  still pass.

janmenjayap committed:
Merge branch 'micali_vazirani' of github.com:janmenjayap/sage into micali_vazirani

dcoudert reviewed (commented):
Let us know when your are done with all the improvements.

janmenjayap committed:
graphs/matching: correct maximum-matching early-exit comments

- Clarify that the `matching_size == N // 2` early exit means the matching is
  maximum, with at most one exposed vertex, rather than necessarily perfect.
  For odd `N`, no perfect matching exists, but the condition can still hold.
- Reword the comment in `get_matching` and add the same note to the matching
  early exit in `MAX`, so the two explanations read consistently.
- Comment-only change; no behavior change.

janmenjayap committed:
graphs/matching: build get_matching result with a generator

- Replace the explicit `Graph` accumulation, which added all vertices and then
  added each matched edge in a loop, with a `matched_edges` generator passed to
  the `Graph` constructor via `format='vertices_and_edges'`, returning its
  `.edges()`.
- Same result: an `EdgesView` over the matched edges on the original vertex
  labels, each edge once. No behavior change.

janmenjayap committed:
graphs/matching_covered_graph: fix a/an grammar in docstrings

- `an 1-regular` -> `a 1-regular` in the module header.
- `a edge` -> `an edge` in the is_matching_covered INPUT block.

janmenjayap committed:
matching_covered_graph: default to the graph's own matching algorithm

- Default `algorithm` to `None` instead of `'Edmonds'` in `__init__`,
  `_upgrade_from_graph`, and `has_perfect_matching`, so each graph uses its own
  default, Micali-Vazirani for a general graph and Hopcroft-Karp for a bipartite
  graph, consistent with the base matching API.
- Accept `None` and `'Micali-Vazirani'` in `has_perfect_matching`'s algorithm
  check; update the error message, docstrings, and the related doctest.
- No result change: a matching-covered graph always has a perfect matching, and
  the construction check is algorithm-independent.

janmenjayap committed:
graphs/matching_covered_graph: thread solver options into the stored matching

  - Pass `algorithm`, `solver`, `verbose` and `integrality_tolerance` from
    `__init__` into the internal `Graph(self).matching()` call instead of
    letting it fall back to the hardcoded defaults, so the stored matching
    honours the algorithm/solver the caller requested.
  - Completes the `algorithm=None` default introduced earlier: the base API's
    default-selection (Micali-Vazirani for an unweighted general graph) now
    actually reaches the stored-matching computation.

janmenjayap committed:
graphs/matching: fix error-message quoting and name the HS2017 variant

  - Correct the malformed quoting in the `algorithm` `ValueError` (and its
    doctest): `'Edmonds', 'LP', or 'Micali-Vazirani'` instead of the misplaced
    commas inside the quotes (`'Edmonds,' 'LP,'`).
  - Note in `search()` that the phase loop implements the *extended search
    phases* variant of Huang and Stein [HS2017]_, matching the class docstring.

janmenjayap committed:
graphs/matching: normalize British spellings in the double DFS

  - Rename `colour` -> `color` and `centre` -> `center` throughout `DDFS`
    (~50 occurrences, including `other_colour`, `red_centre`/`green_centre`,
    `opponent_centre`), matching `self.color` and the rest of the file.
  - No behavior change.

janmenjayap committed:
graphs/matching: use red/green for the double-DFS colouring everywhere

  - Rename `left_support`/`right_support` -> `red_support`/`green_support` at
    the `MAX` unpack, `form_petal`/`_attach_petal_side`, `augment` and its
    `find_path` calls, so the double-DFS output name survives to every consumer.
  - Promote `RED`/`GREEN` from `DDFS`-local constants to class level and pass
    them when attaching petal sides.
  - Replace the `None` sentinel in the per-vertex `color` array with an integer
    `NO_COLOR` (homogeneous `int`, like `INFINITY`/`EXPOSED`), and make the
    even-level unfold test `color[vertex] == RED` explicit instead of relying on
    `not color[vertex]`.
  - No behavior change.

janmenjayap committed:
graphs/matching: clarify misleading identifiers; note the MIN/MAX convention

  - Rename `tenacity_bridges_map` -> `bridges_by_tenacity` (it is a list indexed
    by tenacity, not a map) and `label_max` -> `assign_max_levels`.
  - Document in the class docstring that `MIN`/`MAX`/`DDFS` are uppercased to
    match the procedure names in [MV1980]_/[Vaz2020]_ (a deliberate PEP8
    exception, not to be "fixed").
  - No behavior change.

janmenjayap committed:
graphs/matching: add is_outer()/is_prop()/is_bridge() accessors

  - Add `is_outer(v)` (`even_level < odd_level`) and route the even-level test
    in the petal unfold through it, replacing the cryptic `max_level(v) % 2`.
  - Add `is_prop`/`is_bridge` predicates over `prop_edges` and route the
    bridge test in `assign_max_levels` through `is_bridge`.
  - No behavior change.

  (Vertex `tenacity()` is deferred to the invariant-harness commit, where it
  has a caller; the inline tenacity in MIN/MAX is edge tenacity, a distinct
  quantity.)

janmenjayap committed:
graphs/matching: return a tagged DDFSRun from the double DFS

  - Add a nested `DDFSRun` dataclass (`red_support`, `green_support`,
    `bottleneck`, `encountered_deleted_vertex`, plus an `is_augmenting`
    property) describing the outcome of one `DDFS` run on a bridge.
  - Return it from `DDFS` instead of a positional 4-tuple, and read it by name
    in `MAX`: `run.is_augmenting` replaces the cryptic `bottleneck is None`.
  - No behavior change.

janmenjayap committed:
graphs/matching: surface the HS2017 extended-phase policy as a flag

  - Add an `extended_phases` parameter (default `True`) to the driver, replacing
    the implicit encoding of the Huang-Stein modification (`MAX` returning
    `False` so the phase keeps growing). When `True`, one phase grows and
    augments at successive levels until `MIN` is exhausted; when `False`, the
    phase ends at its first augmenting level (classic Micali-Vazirani).
  - Gate the classic early-exit in `search()` on the flag and correct the
    docstring, which previously mis-described the default as stopping at the
    first augmenting level.
  - Default preserves current behavior; both settings compute a maximum matching.

janmenjayap committed:
graphs/matching: add tenacity() and a check_invariants debug harness

  - Add the vertex accessor `tenacity(v)` = `even_level(v) + odd_level(v)`, the
    central quantity of [Vaz2020]_ (equal to `min_level(v) + max_level(v)`).
  - Add a `check_invariants` constructor keyword (default `False`) and
    `_check_invariants()`: when enabled, `search()` asserts after each level that
    `even_level`/`odd_level` have the right parity and that no vertex is the bud
    of a petal it belongs to, and `assign_max_levels` asserts
    `tenacity(v) == 2*search_level + 1`.
  - Off by default, so no behavior change; the checks are `assert` statements
    (also removed under `python -O`). Enabling them still computes a maximum
    matching, verified by differential fuzzing with the checks on.

janmenjayap committed:
graphs/matching: extract the immutable graph representation into _Instance

  - Move graph simplification (drop loops/multiedges via a plain `Graph`),
    isolated-vertex removal, relabelling to 0..N-1, and edge indexing out of
    `MicaliVaziraniMatching.__init__` into a new nested `_Instance` built once.
  - The driver delegates `edge_to_index` to it and keeps `G`/`N`/
    `index_to_vertex`/`index_to_edge` as references, so the public interface and
    all call sites are unchanged.
  - No behavior change.

janmenjayap committed:
graphs/matching: extract the bud union-find and petals into Petals

  - Move `vertex_bud_map`/`vertex_petal_map`/`color`, `get_bud`, `form_petal`
    and `_attach_petal_side` into a new nested `Petals` unit (the `Petal`
    dataclass is passed into it); `start_new_phase` calls `Petals.reset()`.
  - Route `MAX`, `DDFS` and `find_path` through `self.petals.<method>` directly
    rather than via driver delegators.
  - No behavior change.

janmenjayap committed:
graphs/matching: extract per-phase search state into _SearchState

  - Move the level arrays (`even_level`/`odd_level`), `predecessor`/`successor`,
    `bridges_by_tenacity`, `prop_edges`, `edge_scanned`, `deletion_phase`,
    `search_level_vertices` and the `INFINITY` sentinel into a new nested
    `_SearchState`, together with the derived accessors `min_level`, `max_level`,
    `tenacity`, `is_outer`, `is_prop` and `is_bridge`.
  - The driver constructs `self.state` in `__init__` and accesses all of it
    directly through `self.state.<member>` (no aliases); `start_new_phase`
    resets it in place.
  - No behavior change.

janmenjayap committed:
graphs/matching: drop DDFSRun; return a tuple from the double DFS

  - Remove the `DDFSRun` dataclass: for its single consumer (`MAX`) a tagged
    result object was more machinery than the outcome warranted.
  - `DDFS` now returns a documented tuple `(red_support, green_support,
    bottleneck, encountered_deleted_vertex)`; `bottleneck is None` marks an
    augmenting path, and `MAX` unpacks it into named locals.
  - No behavior change.

janmenjayap committed:
graphs/matching: spell out "double depth-first search"; rename meeting point

  - Write "double depth-first search" in prose rather than "double DFS"; the
    `DDFS` identifier is unchanged.
  - Rename the double-DFS meeting vertex `m` to `meeting_point`.
  - No behavior change.

janmenjayap committed:
graphs/matching: tidy up after the unit extraction

  - Drop the `edge_to_index` delegator on the driver; call
    `self.instance.edge_to_index` directly (matches the get_bud/form_petal
    cleanup).
  - Rename the internal `Petal`/`Petals` classes to `_Petal`/`_Petals` (and
    their type annotations), consistent with `_Instance`/`_SearchState`.
  - Route the `color`/`vertex_petal_map` readers through `self.petals` and drop
    the two driver aliases.
  - No behavior change.

janmenjayap committed:
graphs/matching: remove the dead unfold_path_in_petal wrapper; tidy predicates

    - Remove `unfold_path_in_petal`: it has no caller (all sites use
      `_unfold_path_in_petal_generator` directly via the trampoline).
    - Use `is_exposed` in `_is_valid_augmenting_path` instead of comparing to
      `EXPOSED` directly.
    - Give `_SearchState.max_level` a real caller: `assign_max_levels` now
      asserts (under `check_invariants`) that the level it writes is the larger
      of the two (`[Vaz2020]` Def 3.3), with a doctest that exercises it.
    - Strengthen the `assign_max_levels` doctests to verify a valid maximum
      matching, not just its size.
    - No change to the default (non-debug) behaviour.

janmenjayap committed:
graphs/matching: drop the redundant is_saturated and is_bridge predicates

    - Remove `is_saturated`: it is exactly `not is_exposed`. Keep the
      predicate that mirrors the stored `mate == EXPOSED` state and reroute
      its callers through `not self.is_exposed(...)`.
    - Remove `is_bridge`: it is exactly `not is_prop`. Keep `is_prop`, which
      mirrors the stored `prop_edges` set (there is no `bridge_edges`), and
      reroute its caller through `not self.state.is_prop(...)`.
    - Fold the prop/bridge description into `is_prop`'s docstring.
    - No behavior change.

dcoudert reviewed (commented):
Let me know when you are done with the many rounds of improvements/reshaping of the code.

janmenjayap committed:
Merge branch 'sagemath:develop' into micali_vazirani

janmenjayap committed:
Merge branch 'develop' into micali_vazirani

janmenjayap committed:
graphs/matching: lowercase error messages

- Change "The input must be a graph" to "the input must be a graph" in _Instance.__init__ (dcoudert review comment).
- Change "Micali-Vazirani algorithm does not ..." to "the Micali-Vazirani algorithm does not ..." in matching(), so the message no longer opens with a capitalised word.
- Update the two doctests that assert these exact messages.
- No behavior change.

janmenjayap committed:
graphs/matching: move dataclass and Any imports to module top

- Move `from dataclasses import dataclass` and `from typing import Any` to the module-level imports.
- Avoid leaking `MicaliVaziraniMatching.dataclass` and `MicaliVaziraniMatching.Any` as public class attributes.
- Keep the `@dataclass` decorator on `_Petal` and the `Edge` type alias resolving from module scope.
- No behavior change.

janmenjayap committed:
graphs/matching: make internal state attributes private

- Prefix per-run internal state attributes with underscores:
  - `instance` -> `_instance`
  - `state` -> `_state`
  - `petals` -> `_petals`
  - `mate` -> `_mate`
  - `G` -> `_G`
  - `phase_index` -> `_phase_index`
  - `num_augmentations` -> `_num_augmentations`
  - `matching_size` -> `_matching_size`
- Avoid exposing implementation details as part of the documented API.
- Rename the relabelled internal graph to `_G` to distinguish it from the user's input graph.
- Update doctests to reflect the renamed internal attributes.
- Leave `index_to_vertex`, `index_to_edge`, and `N` unchanged for a follow-up.
- No behavior change.

janmenjayap committed:
graphs/matching: make remaining internal maps and size private

- Rename the internal vertex count from `N` to `_N`, while leaving `_Instance.N` unchanged.
- Remove the driver's `index_to_vertex` and `index_to_edge` aliases.
- Access `index_to_vertex` and `index_to_edge` through `self._instance` at their remaining use sites.
- Match the earlier removal of the `color` and `vertex_petal_map` aliases.
- Update doctests to reflect the renamed and removed internal members.
- Leave only the `extended_phases` and `check_invariants` configuration attributes public.
- No behavior change.

janmenjayap committed:
graphs/matching: make internal phase-step methods private

- Prefix the remaining internal phase-step methods with underscores:
  - `MIN` -> `_MIN`
  - `MAX` -> `_MAX`
  - `DDFS` -> `_DDFS`
  - `search` -> `_search`
  - `augment` -> `_augment`
  - `find_path` -> `_find_path`
  - `unfold_petal` -> `_unfold_petal`
  - `is_exposed` -> `_is_exposed`
  - `start_new_phase` -> `_start_new_phase`
  - `assign_max_levels` -> `_assign_max_levels`
  - `compute_initial_maximal_matching` -> `_compute_initial_maximal_matching`
- Update method definitions, internal call sites, doctests, and `:meth:` cross-references.
- Keep the paper mapping in the docstrings, which still refer to MIN, MAX, and DDFS.
- Leave only `__init__`, `get_matching`, and the `extended_phases`/`check_invariants` configuration attributes as the public interface.
- No behavior change.

janmenjayap committed:
graphs/matching: make sentinel constants and Edge alias private

- Prefix internal sentinel constants with underscores:
  - `EXPOSED` -> `_EXPOSED`
  - `RED` -> `_RED`
  - `GREEN` -> `_GREEN`
  - `NO_COLOR` -> `_NO_COLOR`
- Rename the internal edge type alias from `Edge` to `_Edge`.
- Update definitions, internal constant accesses, `:attr:` and `:class:` cross-references, and the two `_Edge` type annotations.
- Keep the local `RED`/`GREEN` names unpacked in `_DDFS` unchanged.
- Leave only `__init__`, `get_matching`, and the `extended_phases`/`check_invariants` configuration attributes as the public interface.
- No behavior change.

janmenjayap committed:
graphs/matching: refine module and class docstrings

- Rewrite the `MicaliVaziraniMatching` class docstring in a more formal style.
- Introduce `even_level` and `odd_level` before defining `min_level` and `max_level` as their minimum and maximum.
- Correct the earlier conflation of minimum even and maximum odd levels.
- Remove conversational asides and neutralise the naming rationale.
- Split long sentences in the phases and complexity sections for readability.
- Documentation only; no code or doctest changes.

janmenjayap committed:
graphs/matching: doctest classic-phase early exit of _search

- Add a doctest covering the `extended_phases=False` branch of `_search`.
- Exercise the path where a phase augments without completing the matching.
- Use `K_{1,4}` with the greedy seed neutralised so the first phase must augment.
- Cover the early return before `_num_augmentations` is incremented.
- No behavior change.

janmenjayap committed:
Merge branch 'develop' into micali_vazirani
